# Autor: Aaron Frias Romero

# Futbol World Cup Dataset

En este proyecto vamos a trabajar con un conjunto de datos obtenido de [Kaggle](https://www.kaggle.com/abecklas/fifa-world-cup?select=WorldCupMatches.csv) que contiene los datos de la FIFA de la copa del mundo. Utilizaremos estos tres ficheros:

- WorldCupMatches.csv: Datos de los partidos disputados
- WorldCupPlayers.csv: Datos de los jugadores que han jugado en cada partido
- WorldCups.csv: Datos de las copas disputadas y los resultados

# Introducción

Mediante esta libreta vamos a explorar los datos almacenados en estos tres ficheros y a responder preguntas haciendo operaciones y transformaciones con la libreria pandas.

In [131]:
# Importamos la libreria pandas
import pandas as pd

In [132]:
# Cargamos el conjunto de datos WorldCupMatches
df_partidos = pd.read_csv("WorldCupMatches.csv")

## 1. Exploración


Hacemos una primera visualización de los datos para conocer sus columnas y valores.

In [133]:
df_partidos.head()

Year              Datetime    Stage         Stadium         City  \
0  1930.0  13 Jul 1930 - 15:00   Group 1         Pocitos  Montevideo    
1  1930.0  13 Jul 1930 - 15:00   Group 4  Parque Central  Montevideo    
2  1930.0  14 Jul 1930 - 12:45   Group 2  Parque Central  Montevideo    
3  1930.0  14 Jul 1930 - 14:50   Group 3         Pocitos  Montevideo    
4  1930.0  15 Jul 1930 - 16:00   Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France              4.0              1.0         Mexico   
1            USA              3.0              0.0        Belgium   
2     Yugoslavia              2.0              1.0         Brazil   
3        Romania              3.0              1.0           Peru   
4      Argentina              1.0              0.0         France   

  Win conditions  Attendance  Half-time Home Goals  Half-time Away Goals  \
0                     4444.0                   3.0                   0.0   
1                    18346.0                   2.0                   0.0   
2                    24059.0                   2.0                   0.0   
3                     2549.0                   1.0                   0.0   
4                    23409.0                   0.0                   0.0   

                  Referee               Assistant 1  \
0  LOMBARDI Domingo (URU)     CRISTOPHE Henry (BEL)   
1       MACIAS Jose (ARG)  MATEUCCI Francisco (URU)   
2     TEJADA Anibal (URU)   VALLARINO Ricardo (URU)   
3   WARNKEN Alberto (CHI)       LANGENUS Jean (BEL)   
4     REGO Gilberto (BRA)      SAUCEDO Ulises (BOL)   

                  Assistant 2  RoundID  MatchID Home Team Initials  \
0         REGO Gilberto (BRA)    201.0   1096.0                FRA   
1       WARNKEN Alberto (CHI)    201.0   1090.0                USA   
2         BALWAY Thomas (FRA)    201.0   1093.0                YUG   
3    MATEUCCI Francisco (URU)    201.0   1098.0                ROU   
4  RADULESCU Constantin (ROU)    201.0   1085.0                ARG   

  Away Team Initials  
0                MEX  
1                BEL  
2                BRA  
3                PER  
4                FRA

Ahora que hemos hecho una primera aproximación, observamos la tipologia de las variables con las que vamos a trabajar.

In [134]:
df_partidos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4572 entries, 0 to 4571
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Year                  852 non-null    float64
 1   Datetime              852 non-null    object 
 2   Stage                 852 non-null    object 
 3   Stadium               852 non-null    object 
 4   City                  852 non-null    object 
 5   Home Team Name        852 non-null    object 
 6   Home Team Goals       852 non-null    float64
 7   Away Team Goals       852 non-null    float64
 8   Away Team Name        852 non-null    object 
 9   Win conditions        852 non-null    object 
 10  Attendance            850 non-null    float64
 11  Half-time Home Goals  852 non-null    float64
 12  Half-time Away Goals  852 non-null    float64
 13  Referee               852 non-null    object 
 14  Assistant 1           852 non-null    object 
 15  Assistant 2          

In [135]:
df_partidos.describe().transpose()

count          mean           std     min       25%  \
Year                  852.0  1.985089e+03  2.244882e+01  1930.0   1970.00   
Home Team Goals       852.0  1.811033e+00  1.610255e+00     0.0      1.00   
Away Team Goals       852.0  1.022300e+00  1.087573e+00     0.0      0.00   
Attendance            850.0  4.516480e+04  2.348525e+04  2000.0  30000.00   
Half-time Home Goals  852.0  7.089202e-01  9.374141e-01     0.0      0.00   
Half-time Away Goals  852.0  4.284038e-01  6.912519e-01     0.0      0.00   
RoundID               852.0  1.066177e+07  2.729613e+07   201.0    262.00   
MatchID               852.0  6.134687e+07  1.110572e+08    25.0   1188.75   

                          50%          75%          max  
Year                   1990.0      2002.00       2014.0  
Home Team Goals           2.0         3.00         10.0  
Away Team Goals           1.0         2.00          7.0  
Attendance            41579.5     61374.50     173850.0  
Half-time Home Goals      0.0         1.00          6.0  
Half-time Away Goals      0.0         1.00          5.0  
RoundID                 337.0    249722.00   97410600.0  
MatchID                2191.0  43950059.25  300186515.0

In [136]:
df_partidos.shape

(4572, 20)

In [137]:
df_partidos.size

91440

Podemos observar que disponemos de **4572 filas y 20 columnas.**
Disponemos de variables que se corresponden con la fecha de los partidos:

*   Year
*   Datetime

Columnas relacionadas con la localidad dónde se jugó el partido:

*   Stage
*   Stadium
*   City

Y otras variables relacionadas con el partido en sí, como los equipos implicados (Home Team Name y Away Team Name), sus iniciales (Home Team Name y Away Team Name) y el resultado del partido (Home Team Goals y Away Team Goals).

Un dato relevante es que vemos que disponemos de 4572 filas, pero tan solo un conteo de 852 partidos, es decir, 3720 filas con valores perdidos.

## 2. Limpieza de datos

Una de las cosas que llama la atención es la alta presencia de valores perdidos en el DataFrame `df_partidos`.

Para ello vamos a eliminar estos valores perdidos del conjunto del dataset.

In [138]:
# Observamos los valores nulos
df_partidos.isnull().sum()

Year                    3720
Datetime                3720
Stage                   3720
Stadium                 3720
City                    3720
Home Team Name          3720
Home Team Goals         3720
Away Team Goals         3720
Away Team Name          3720
Win conditions          3720
Attendance              3722
Half-time Home Goals    3720
Half-time Away Goals    3720
Referee                 3720
Assistant 1             3720
Assistant 2             3720
RoundID                 3720
MatchID                 3720
Home Team Initials      3720
Away Team Initials      3720
dtype: int64

Observamos que hay 3720 valores null en todas las columnas.
Para eliminarlas y limpiar el df realizaremos lo siguiente.

In [139]:
# Eliminamos las filas con algún valor perdido
df_partidos = df_partidos.dropna()
df_partidos

Year              Datetime                     Stage  \
0    1930.0  13 Jul 1930 - 15:00                    Group 1   
1    1930.0  13 Jul 1930 - 15:00                    Group 4   
2    1930.0  14 Jul 1930 - 12:45                    Group 2   
3    1930.0  14 Jul 1930 - 14:50                    Group 3   
4    1930.0  15 Jul 1930 - 16:00                    Group 1   
..      ...                   ...                       ...   
847  2014.0  05 Jul 2014 - 17:00             Quarter-finals   
848  2014.0  08 Jul 2014 - 17:00                Semi-finals   
849  2014.0  09 Jul 2014 - 17:00                Semi-finals   
850  2014.0  12 Jul 2014 - 17:00   Play-off for third place   
851  2014.0  13 Jul 2014 - 16:00                      Final   

                 Stadium             City Home Team Name  Home Team Goals  \
0                Pocitos      Montevideo          France              4.0   
1         Parque Central      Montevideo             USA              3.0   
2         Parque Central      Montevideo      Yugoslavia              2.0   
3                Pocitos      Montevideo         Romania              3.0   
4         Parque Central      Montevideo       Argentina              1.0   
..                   ...              ...            ...              ...   
847     Arena Fonte Nova        Salvador     Netherlands              0.0   
848     Estadio Mineirao  Belo Horizonte          Brazil              1.0   
849   Arena de Sao Paulo       Sao Paulo     Netherlands              0.0   
850     Estadio Nacional        Brasilia          Brazil              0.0   
851  Estadio do Maracana  Rio De Janeiro         Germany              1.0   

     Away Team Goals Away Team Name                         Win conditions  \
0                1.0         Mexico                                          
1                0.0        Belgium                                          
2                1.0         Brazil                                          
3                1.0           Peru                                          
4                0.0         France                                          
..               ...            ...                                    ...   
847              0.0     Costa Rica  Netherlands win on penalties (4 - 3)    
848              7.0        Germany                                          
849              0.0      Argentina    Argentina win on penalties (2 - 4)    
850              3.0    Netherlands                                          
851              0.0      Argentina          Germany win after extra time    

     Attendance  Half-time Home Goals  Half-time Away Goals  \
0        4444.0                   3.0                   0.0   
1       18346.0                   2.0                   0.0   
2       24059.0                   2.0                   0.0   
3        2549.0                   1.0                   0.0   
4       23409.0                   0.0                   0.0   
..          ...                   ...                   ...   
847     51179.0                   0.0                   0.0   
848     58141.0                   0.0                   5.0   
849     63267.0                   0.0                   0.0   
850     68034.0                   0.0                   2.0   
851     74738.0                   0.0                   0.0   

                    Referee                  Assistant 1  \
0    LOMBARDI Domingo (URU)        CRISTOPHE Henry (BEL)   
1         MACIAS Jose (ARG)     MATEUCCI Francisco (URU)   
2       TEJADA Anibal (URU)      VALLARINO Ricardo (URU)   
3     WARNKEN Alberto (CHI)          LANGENUS Jean (BEL)   
4       REGO Gilberto (BRA)         SAUCEDO Ulises (BOL)   
..                      ...                          ...   
847   Ravshan IRMATOV (UZB)  RASULOV Abduxamidullo (UZB)   
848   RODRIGUEZ Marco (MEX)      TORRENTERA Marvin (MEX)   
849      C�neyt �AKIR (TUR)         DURAN Bahattin (TUR)   
850   HAIMOUDI Djamel (ALG)    

## 3. Cargar datos

Al igual que hemos cargado los datos de los partidos en `df_partidos`, creamos las variables `df_jugadores` y `df_copas` que contengan los dataframes correspondientes a la lectura de los csv `WorldCupPlayers` y `WorldCups`.


In [140]:
df_jugadores = pd.read_csv("WorldCupPlayers.csv")
df_jugadores.head()

RoundID  MatchID Team Initials           Coach Name Line-up  Shirt Number  \
0      201     1096           FRA  CAUDRON Raoul (FRA)       S             0   
1      201     1096           MEX     LUQUE Juan (MEX)       S             0   
2      201     1096           FRA  CAUDRON Raoul (FRA)       S             0   
3      201     1096           MEX     LUQUE Juan (MEX)       S             0   
4      201     1096           FRA  CAUDRON Raoul (FRA)       S             0   

        Player Name Position Event  
0       Alex THEPOT       GK   NaN  
1   Oscar BONFIGLIO       GK   NaN  
2  Marcel LANGILLER      NaN  G40'  
3      Juan CARRENO      NaN  G70'  
4   Ernest LIBERATI      NaN   NaN

In [141]:
df_copas = pd.read_csv("WorldCups.csv")
df_copas.head()

Year      Country      Winner      Runners-Up    Third      Fourth  \
0  1930      Uruguay     Uruguay       Argentina      USA  Yugoslavia   
1  1934        Italy       Italy  Czechoslovakia  Germany     Austria   
2  1938       France       Italy         Hungary   Brazil      Sweden   
3  1950       Brazil     Uruguay          Brazil   Sweden       Spain   
4  1954  Switzerland  Germany FR         Hungary  Austria     Uruguay   

   GoalsScored  QualifiedTeams  MatchesPlayed Attendance  
0           70              13             18    590.549  
1           70              16             17    363.000  
2           84              15             18    375.700  
3           88              13             22  1.045.246  
4          140              16             26    768.607

---
# Conclusiones y preguntas a responder

---

## 4. Rango temporal

El dataframe `df_copas` contiene datos de todos los mundiales disputados. Podemos responder a distintas preguntas:
*  ¿Cuál es el año del mundial más antiguo disputado?
*  ¿Y el año del mundial más reciente?

In [142]:
print(f"El mundial más antiguo se jugó en el año {df_copas.Year.min()}, y el más reciente en el año {df_copas.Year.max()}")

El mundial más antiguo se jugó en el año 1930, y el más reciente en el año 2014


In [143]:
df_copas.loc[df_copas.Year == df_copas.Year.min()]

Year  Country   Winner Runners-Up Third      Fourth  GoalsScored  \
0  1930  Uruguay  Uruguay  Argentina   USA  Yugoslavia           70   

   QualifiedTeams  MatchesPlayed Attendance  
0              13             18    590.549

In [144]:
df_copas.loc[df_copas.Year == df_copas.Year.max()]

Year Country   Winner Runners-Up        Third  Fourth  GoalsScored  \
19  2014  Brazil  Germany  Argentina  Netherlands  Brazil          171   

    QualifiedTeams  MatchesPlayed Attendance  
19              32             64  3.386.810

## 5. Asistencia de público

La columna `Attendance` indica la asistencia de público en cada mundial. Pandas ha cargado esta columna como tipo string, ya que los valores indicados contienen puntos como separador de millares. En primer lugar vamos a transformar esta columna para que no contenga el caracter `.`. Posteriormente transformaremos esta columna a formato numérico.

In [145]:
df_copas = pd.read_csv("WorldCups.csv", thousands=".")

Teniendo en cuenta todos los datos de `df_copas`, ¿Cuál fué la mayor asistencia de un mundial?

In [146]:
df_copas.head()

Year      Country      Winner      Runners-Up    Third      Fourth  \
0  1930      Uruguay     Uruguay       Argentina      USA  Yugoslavia   
1  1934        Italy       Italy  Czechoslovakia  Germany     Austria   
2  1938       France       Italy         Hungary   Brazil      Sweden   
3  1950       Brazil     Uruguay          Brazil   Sweden       Spain   
4  1954  Switzerland  Germany FR         Hungary  Austria     Uruguay   

   GoalsScored  QualifiedTeams  MatchesPlayed  Attendance  
0           70              13             18      590549  
1           70              16             17      363000  
2           84              15             18      375700  
3           88              13             22     1045246  
4          140              16             26      768607

In [147]:
df_copas.agg(["max","min"])

Year    Country     Winner Runners-Up    Third      Fourth  GoalsScored  \
max  2014    Uruguay    Uruguay     Sweden      USA  Yugoslavia          171   
min  1930  Argentina  Argentina  Argentina  Austria     Austria           70   

     QualifiedTeams  MatchesPlayed  Attendance  
max              32             64     3587538  
min              13             17      363000

In [148]:
df_copas['Attendance'].max()

3587538

In [149]:
df_copas['Attendance'].min()

363000

## 6. Asistencia de público en función del país

Sabemos que, dependiendo del país donde se celebre la copa del mundo, ésta tiene un mayor o menor impacto o relevancia. Vamos a calcular la media de asistencia agrupando por país. Finalmente, ordenar los resultados por orden decreciente de este número medio de asistentes.


In [150]:
Attend = df_copas.groupby("Country").mean().Attendance
Attend

<ipython-input-150-2f1dabe2f227>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  Attend = df_copas.groupby("Country").mean().Attendance


Country
Argentina       1545791.0
Brazil          2216028.0
Chile            893172.0
England         1563135.0
France          1580400.0
Germany         2612596.0
Italy           1439607.5
Korea/Japan     2705197.0
Mexico          1999003.0
South Africa    3178856.0
Spain           2109723.0
Sweden           819810.0
Switzerland      768607.0
USA             3587538.0
Uruguay          590549.0
Name: Attendance, dtype: float64

In [151]:
Attend.sort_values(ascending= False)

Country
USA             3587538.0
South Africa    3178856.0
Korea/Japan     2705197.0
Germany         2612596.0
Brazil          2216028.0
Spain           2109723.0
Mexico          1999003.0
France          1580400.0
England         1563135.0
Argentina       1545791.0
Italy           1439607.5
Chile            893172.0
Sweden           819810.0
Switzerland      768607.0
Uruguay          590549.0
Name: Attendance, dtype: float64

## 7. País con más victorias

Vamos a analizar el número de veces que un país ha ganado la copa del mundo. Transformar el DataFrame `df_copas` para que muestre los diferentes países que han ganado al menos una copa del mundo, y ordena el DataFrame por dicho número de forma descendente usando la función [sort_values](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html).

¿Ves algo extraño o digno de comentar? Pista: analiza Alemania. **Hazlo en una nueva celda de tipo markdown.**

In [152]:
df_copas["Winner"].value_counts().sort_values(ascending = False)

Brazil        5
Italy         4
Germany FR    3
Uruguay       2
Argentina     2
England       1
France        1
Spain         1
Germany       1
Name: Winner, dtype: int64

Como observamos Alemania (Germany) aparece dos veces.

Esto se deberá a que la última copa del mundo ganada Alemania ya era un país unificado, antes era una República Federal. Es por eso que existen dos valores de Alemania en este dataset.

## Creación de un nuevo DataFrame

A continuación vamos a trabajar con el DataFrame `df_partidos`. Vamos a centrarnos en la información a nivel de selección, siendo indiferente si el equipo juega en casa o fuera. En este sentido, la información disponible no es la mejor para trabajar cómodamente. Por ello, es habitual realizar transformaciones a los datos en función de las preguntas que le queramos hacer. A continuación crearemos un nuevo DataFrame llamado `df_paises`que contendrá todos los paises involucrados en la copa del mundo y el año en el que participaron. Es decir, un país que haya jugado varias copas del mundo aparecerá varias veces:


||Year|Country|
|---|---|---|
|0|1930|France|
|1|1930|USA|
|...|...|..|
|37|1938|France|
|...|...|..|


In [153]:
#alternativa 1
df_paises = (
    pd.concat([
        df_partidos[["Year", "Home Team Name"]].rename(columns={"Home Team Name": "Country"}),
        df_partidos[["Year", "Away Team Name"]].rename(columns={"Away Team Name": "Country"})
    ])
)
#alternativa 2
df_paises = (
    df_partidos
    .melt(id_vars=["Year"], value_vars=["Home Team Name", "Away Team Name"])
    .rename(columns={"value": "Country"})
    [["Year", "Country"]]
)

In [154]:
df_paises = df_paises.dropna()
df_paises

Year      Country
0     1930.0       France
1     1930.0          USA
2     1930.0   Yugoslavia
3     1930.0      Romania
4     1930.0    Argentina
...      ...          ...
1695  2014.0   Costa Rica
1696  2014.0      Germany
1697  2014.0    Argentina
1698  2014.0  Netherlands
1699  2014.0    Argentina

[1700 rows x 2 columns]

## 8. Número de países

Vamos a descubrir cuántos países han participado, al menos una vez, en una copa del mundo.

In [155]:
df_paises["Country"].nunique()

83

Con esta función observamos que hay 83 países que han participado al menos 1 vez en una copa del mundo de un total de 20 copas del mundo.

## 9. Número de partidos por país

Ahora vamos a descubrir cuántos partidos ha jugado cada país.

In [156]:
df_paises["Country"].value_counts().sort_values(ascending = False)

Brazil                        108
Italy                          83
Argentina                      81
Germany FR                     62
England                        62
                             ... 
Haiti                           3
rn">Bosnia and Herzegovina      3
Israel                          3
Iraq                            3
Dutch East Indies               1
Name: Country, Length: 83, dtype: int64

## 10. Partidos jugados por España

Vamos a trabajar a continuación con el DataFrame `df_partidos`. Analizaremos aquellos partidos (filas del dataframe) en los que España ha jugado como `Home Team`, es decir, donde la columna `Home Team` tiene el valor `España`.

A continuación, haremos una nueva selección en los que España ha jugado como `Away Team`, es decir, donde la columna `Away Team` tiene el valor `España`.


In [157]:
df_partidos = df_partidos.dropna()
df_partidos.sample(5)

Year              Datetime                  Stage  \
358  1982.0  10 Jul 1982 - 20:00   Match for third place   
641  2002.0  26 Jun 2002 - 20:30             Semi-finals   
544  1998.0  21 Jun 1998 - 17:30                 Group H   
159  1962.0  07 Jun 1962 - 15:00                 Group 2   
402  1986.0  18 Jun 1986 - 12:00             Round of 16   

                  Stadium                City Home Team Name  Home Team Goals  \
358       Jose Rico Perez           Alicante          Poland              3.0   
641  Saitama Stadium 2002            Saitama          Brazil              1.0   
544      Parc des Princes              Paris       Argentina              5.0   
159              Nacional  Santiago De Chile           Italy              3.0   
402        Estadio Azteca        Mexico City         England              3.0   

     Away Team Goals Away Team Name Win conditions  Attendance  \
358              2.0         France                    28000.0   
641              0.0         Turkey                    61058.0   
544              0.0        Jamaica                    45500.0   
159              0.0    Switzerland                    59828.0   
402              0.0       Paraguay                    98728.0   

     Half-time Home Goals  Half-time Away Goals                   Referee  \
358                   2.0                   1.0     GARRIDO Antonio (POR)   
641                   0.0                   0.0  NIELSEN Kim Milton (DEN)   
544                   1.0                   0.0       PEDERSEN Rune (NOR)   
159                   1.0                   0.0    LATYCHEV Nikolaj (URS)   
402                   1.0                   0.0     AL SHARIF Jamal (SYR)   

                   Assistant 1                    Assistant 2     RoundID  \
358  RUBIO VAZQUEZ Mario (MEX)           LACARNE Belaid (ALG)       676.0   
641   WIERZBOWSKI Maciej (POL)              SRAMKA Igor (SVK)  43950400.0   
544       NILSSON Mikael (SWE)           MAZZEI Gennaro (ITA)      1014.0   
159         DAVIDSON Bob (SCO)       RUMENTCHEV Dimitar (BUL)       231.0   
402        PONNET Alexis (BEL)  AL SHANAR Fallaj Khuzam (KSA)       309.0   

        MatchID Home Team Initials Away Team Initials  
358       921.0                POL                FRA  
641  43950062.0                BRA                TUR  
544      8752.0                ARG                JAM  
159      1532.0                ITA                SUI  
402       536.0                ENG                PAR

In [158]:
df_partidos.set_index("Home Team Name").loc["Spain"]

Year               Datetime              Stage  \
Home Team Name                                                     
Spain           1934.0   27 May 1934 - 16:30   Preliminary round   
Spain           1950.0   25 Jun 1950 - 15:00             Group 2   
Spain           1950.0   29 Jun 1950 - 15:00             Group 2   
Spain           1950.0   02 Jul 1950 - 15:00             Group 2   
Spain           1962.0   03 Jun 1962 - 15:00             Group 3   
Spain           1966.0   15 Jul 1966 - 19:30             Group 2   
Spain           1978.0   11 Jun 1978 - 13:45             Group 3   
Spain           1982.0   16 Jun 1982 - 21:00             Group 5   
Spain           1982.0   20 Jun 1982 - 21:00             Group 5   
Spain           1982.0   05 Jul 1982 - 21:00             Group 2   
Spain           1986.0   01 Jun 1986 - 12:00             Group D   
Spain           1986.0   22 Jun 1986 - 16:00      Quarter-finals   
Spain           1990.0   26 Jun 1990 - 17:00         Round of 16   
Spain           1994.0   17 Jun 1994 - 19:30             Group C   
Spain           1994.0   02 Jul 1994 - 16:30         Round of 16   
Spain           1998.0   13 Jun 1998 - 14:30             Group D   
Spain           1998.0   19 Jun 1998 - 21:00             Group D   
Spain           1998.0   24 Jun 1998 - 21:00             Group D   
Spain           2002.0   02 Jun 2002 - 20:30             Group B   
Spain           2002.0  07 June 2002 - 18:00             Group B   
Spain           2002.0   16 Jun 2002 - 20:30         Round of 16   
Spain           2002.0   22 Jun 2002 - 15:30      Quarter-finals   
Spain           2006.0   14 Jun 2006 - 15:00             Group H   
Spain           2006.0   19 Jun 2006 - 21:00             Group H   
Spain           2006.0   27 Jun 2006 - 21:00         Round of 16   
Spain           2010.0   16 Jun 2010 - 16:00             Group H   
Spain           2010.0   21 Jun 2010 - 20:30             Group H   
Spain           2010.0   29 Jun 2010 - 20:30         Round of 16   
Spain           2014.0   13 Jun 2014 - 16:00             Group B   
Spain           2014.0   18 Jun 2014 - 16:00             Group B   

                                                  Stadium             City  \
Home Team Name                                                               
Spain                                      Luigi Ferraris           Genoa    
Spain                                    Durival de Brito        Curitiba    
Spain           Maracan� - Est�dio Jornalista M�rio Filho  Rio De Janeiro    
Spain           Maracan� - Est�dio Jornalista M�rio Filho  Rio De Janeiro    
Spain                                   Estadio Sausalito    Vina Del Mar    
Spain                                        Hillsborough       Sheffield    
Spain                                     Jose Amalfitani    Buenos Aires    
Spain                                       Luis Casanova        Valencia    
Spain                                       Luis Casanova        Valencia    
Spain                                   Santiago Bernabeu          Madrid    
Spain                                             Jalisco     Guadalajara    
Spain                                          Cuauhtemoc          Puebla    
Spain                              Marc Antonio Bentegodi          Verona    
Spain                                         Cotton Bowl          Dallas    
Spain                                         RFK Stadium   Washington Dc    
Spain                                        La Beaujoire          Nantes    
Spain                             Stade Geoffroy Guichard   Saint-Etienne    
Spain                                Stade Felix Bollaert            Lens    
Spain                           Gwangju World Cup Stadium         Gwangju    
Spain                            Jeonju World Cup Stadium          Jeonju    
Spain                             Suwon World Cup Stadium           Suwon    
Spain                           Gwangju World C

In [159]:
df_partidos.set_index("Away Team Name").loc["Spain"]

Year              Datetime           Stage  \
Away Team Name                                                 
Spain           1934.0  31 May 1934 - 16:30   Quarter-finals   
Spain           1934.0  01 Jun 1934 - 16:30   Quarter-finals   
Spain           1950.0  09 Jul 1950 - 15:00          Group 6   
Spain           1950.0  13 Jul 1950 - 15:00          Group 6   
Spain           1950.0  16 Jul 1950 - 15:00          Group 6   
Spain           1962.0  31 May 1962 - 15:00          Group 3   
Spain           1962.0  06 Jun 1962 - 15:00          Group 3   
Spain           1966.0  13 Jul 1966 - 19:30          Group 2   
Spain           1966.0  20 Jul 1966 - 19:30          Group 2   
Spain           1978.0  03 Jun 1978 - 13:45          Group 3   
Spain           1978.0  07 Jun 1978 - 13:45          Group 3   
Spain           1982.0  25 Jun 1982 - 21:00          Group 5   
Spain           1982.0  02 Jul 1982 - 21:00          Group 2   
Spain           1986.0  07 Jun 1986 - 12:00          Group D   
Spain           1986.0  12 Jun 1986 - 12:00          Group D   
Spain           1986.0  18 Jun 1986 - 16:00      Round of 16   
Spain           1990.0  13 Jun 1990 - 17:00          Group E   
Spain           1990.0  17 Jun 1990 - 21:00          Group E   
Spain           1990.0  21 Jun 1990 - 17:00          Group E   
Spain           1994.0  21 Jun 1994 - 16:00          Group C   
Spain           1994.0  27 Jun 1994 - 16:00          Group C   
Spain           1994.0  09 Jul 1994 - 12:00   Quarter-finals   
Spain           2002.0  12 Jun 2002 - 20:30          Group B   
Spain           2006.0  23 Jun 2006 - 16:00          Group H   
Spain           2010.0  25 Jun 2010 - 20:30          Group H   
Spain           2010.0  03 Jul 2010 - 20:30   Quarter-finals   
Spain           2010.0  07 Jul 2010 - 20:30      Semi-finals   
Spain           2010.0  11 Jul 2010 - 20:30            Final   
Spain           2014.0  23 Jun 2014 - 13:00          Group B   

                                                  Stadium               City  \
Away Team Name                                                                 
Spain                                      Giovanni Berta          Florence    
Spain                                      Giovanni Berta          Florence    
Spain                                            Pacaembu         Sao Paulo    
Spain           Maracan� - Est�dio Jornalista M�rio Filho    Rio De Janeiro    
Spain                                            Pacaembu         Sao Paulo    
Spain                                   Estadio Sausalito      Vina Del Mar    
Spain                                   Estadio Sausalito      Vina Del Mar    
Spain                                          Villa Park        Birmingham    
Spain                                          Villa Park        Birmingham    
Spain                                     Jose Amalfitani      Buenos Aires    
Spain                          Estadio Jos� Mar�a Minella     Mar Del Plata    
Spain                                       Luis Casanova          Valencia    
Spain                                   Santiago Bernabeu            Madrid    
Spain                                       Tres de Marzo       Guadalajara    
Spain                                         Tecnologico         Monterrey    
Spain                                 Estadio Corregidora         Queretaro    
Spain                                         Dacia Arena             Udine    
Spain                                         Dacia Arena             Udine    
Spain                              Marc Antonio Bentegodi            Verona    
Spain                                       Soldier Field           Chicago    
Spain                                       Soldier Field           Chicago    
Spain                                     Foxboro Stadium            Boston    
Spain                           Daejeon World Cup Stadium           Daejeon    
Spain                            

In [160]:
df_partidos.set_index("Home Team Name","Away Team Name").loc["Spain"]

<ipython-input-160-758eaa8bfdf2>:1: FutureWarning: In a future version of pandas all arguments of DataFrame.set_index except for the argument 'keys' will be keyword-only.
  df_partidos.set_index("Home Team Name","Away Team Name").loc["Spain"]


Year               Datetime              Stage  \
Home Team Name                                                     
Spain           1934.0   27 May 1934 - 16:30   Preliminary round   
Spain           1950.0   25 Jun 1950 - 15:00             Group 2   
Spain           1950.0   29 Jun 1950 - 15:00             Group 2   
Spain           1950.0   02 Jul 1950 - 15:00             Group 2   
Spain           1962.0   03 Jun 1962 - 15:00             Group 3   
Spain           1966.0   15 Jul 1966 - 19:30             Group 2   
Spain           1978.0   11 Jun 1978 - 13:45             Group 3   
Spain           1982.0   16 Jun 1982 - 21:00             Group 5   
Spain           1982.0   20 Jun 1982 - 21:00             Group 5   
Spain           1982.0   05 Jul 1982 - 21:00             Group 2   
Spain           1986.0   01 Jun 1986 - 12:00             Group D   
Spain           1986.0   22 Jun 1986 - 16:00      Quarter-finals   
Spain           1990.0   26 Jun 1990 - 17:00         Round of 16   
Spain           1994.0   17 Jun 1994 - 19:30             Group C   
Spain           1994.0   02 Jul 1994 - 16:30         Round of 16   
Spain           1998.0   13 Jun 1998 - 14:30             Group D   
Spain           1998.0   19 Jun 1998 - 21:00             Group D   
Spain           1998.0   24 Jun 1998 - 21:00             Group D   
Spain           2002.0   02 Jun 2002 - 20:30             Group B   
Spain           2002.0  07 June 2002 - 18:00             Group B   
Spain           2002.0   16 Jun 2002 - 20:30         Round of 16   
Spain           2002.0   22 Jun 2002 - 15:30      Quarter-finals   
Spain           2006.0   14 Jun 2006 - 15:00             Group H   
Spain           2006.0   19 Jun 2006 - 21:00             Group H   
Spain           2006.0   27 Jun 2006 - 21:00         Round of 16   
Spain           2010.0   16 Jun 2010 - 16:00             Group H   
Spain           2010.0   21 Jun 2010 - 20:30             Group H   
Spain           2010.0   29 Jun 2010 - 20:30         Round of 16   
Spain           2014.0   13 Jun 2014 - 16:00             Group B   
Spain           2014.0   18 Jun 2014 - 16:00             Group B   

                                                  Stadium             City  \
Home Team Name                                                               
Spain                                      Luigi Ferraris           Genoa    
Spain                                    Durival de Brito        Curitiba    
Spain           Maracan� - Est�dio Jornalista M�rio Filho  Rio De Janeiro    
Spain           Maracan� - Est�dio Jornalista M�rio Filho  Rio De Janeiro    
Spain                                   Estadio Sausalito    Vina Del Mar    
Spain                                        Hillsborough       Sheffield    
Spain                                     Jose Amalfitani    Buenos Aires    
Spain                                       Luis Casanova        Valencia    
Spain                                       Luis Casanova        Valencia    
Spain                                   Santiago Bernabeu          Madrid    
Spain                                             Jalisco     Guadalajara    
Spain                                          Cuauhtemoc          Puebla    
Spain                              Marc Antonio Bentegodi          Verona    
Spain                                         Cotton Bowl          Dallas    
Spain                                         RFK Stadium   Washington Dc    
Spain                                        La Beaujoire          Nantes    
Spain                             Stade Geoffroy Guichard   Saint-Etienne    
Spain                                Stade Felix Bollaert            Lens    
Spain                           Gwangju World Cup Stadium         Gwangju    
Spain                            Jeonju World Cup Stadium          Jeonju    
Spain                             Suwon World Cup Stadium           Suwon    
Spain                           Gwangju World C

## Calcular el vencedor de cada partido

El DataFrame `df_partidos` tiene la información de cada partido, incluyendo los goles marcados por cada equipo, pero no disponemos de una columna que indique si un equipo ganó o perdió. Vamos a crear una columna llamada `Ganador` que indique el nombre del equipo vencedor (con más goles).

In [161]:
df_partidos["Ganador"] = df_partidos["Home Team Name"]
df_partidos.loc[df_partidos["Home Team Goals"] < df_partidos["Away Team Goals"], "Ganador"] = df_partidos["Away Team Name"]
df_partidos.loc[df_partidos["Home Team Goals"] == df_partidos["Away Team Goals"], "Ganador"] = pd.NA
df_partidos.head()

Year              Datetime    Stage         Stadium         City  \
0  1930.0  13 Jul 1930 - 15:00   Group 1         Pocitos  Montevideo    
1  1930.0  13 Jul 1930 - 15:00   Group 4  Parque Central  Montevideo    
2  1930.0  14 Jul 1930 - 12:45   Group 2  Parque Central  Montevideo    
3  1930.0  14 Jul 1930 - 14:50   Group 3         Pocitos  Montevideo    
4  1930.0  15 Jul 1930 - 16:00   Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France              4.0              1.0         Mexico   
1            USA              3.0              0.0        Belgium   
2     Yugoslavia              2.0              1.0         Brazil   
3        Romania              3.0              1.0           Peru   
4      Argentina              1.0              0.0         France   

  Win conditions  ...  Half-time Home Goals  Half-time Away Goals  \
0                 ...                   3.0                   0.0   
1                 ...                   2.0                   0.0   
2                 ...                   2.0                   0.0   
3                 ...                   1.0                   0.0   
4                 ...                   0.0                   0.0   

                  Referee               Assistant 1  \
0  LOMBARDI Domingo (URU)     CRISTOPHE Henry (BEL)   
1       MACIAS Jose (ARG)  MATEUCCI Francisco (URU)   
2     TEJADA Anibal (URU)   VALLARINO Ricardo (URU)   
3   WARNKEN Alberto (CHI)       LANGENUS Jean (BEL)   
4     REGO Gilberto (BRA)      SAUCEDO Ulises (BOL)   

                  Assistant 2 RoundID  MatchID  Home Team Initials  \
0         REGO Gilberto (BRA)   201.0   1096.0                 FRA   
1       WARNKEN Alberto (CHI)   201.0   1090.0                 USA   
2         BALWAY Thomas (FRA)   201.0   1093.0                 YUG   
3    MATEUCCI Francisco (URU)   201.0   1098.0                 ROU   
4  RADULESCU Constantin (ROU)   201.0   1085.0                 ARG   

  Away Team Initials     Ganador  
0                MEX      France  
1                BEL         USA  
2                BRA  Yugoslavia  
3                PER     Romania  
4                FRA   Argentina  

[5 rows x 21 columns]

## 11. Número de victorias por país

Calcularemos para cada país el número de partidos ganados.

In [162]:
df_partidos = df_partidos.dropna()
df_partidos.head()

Year              Datetime    Stage         Stadium         City  \
0  1930.0  13 Jul 1930 - 15:00   Group 1         Pocitos  Montevideo    
1  1930.0  13 Jul 1930 - 15:00   Group 4  Parque Central  Montevideo    
2  1930.0  14 Jul 1930 - 12:45   Group 2  Parque Central  Montevideo    
3  1930.0  14 Jul 1930 - 14:50   Group 3         Pocitos  Montevideo    
4  1930.0  15 Jul 1930 - 16:00   Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France              4.0              1.0         Mexico   
1            USA              3.0              0.0        Belgium   
2     Yugoslavia              2.0              1.0         Brazil   
3        Romania              3.0              1.0           Peru   
4      Argentina              1.0              0.0         France   

  Win conditions  ...  Half-time Home Goals  Half-time Away Goals  \
0                 ...                   3.0                   0.0   
1                 ...                   2.0                   0.0   
2                 ...                   2.0                   0.0   
3                 ...                   1.0                   0.0   
4                 ...                   0.0                   0.0   

                  Referee               Assistant 1  \
0  LOMBARDI Domingo (URU)     CRISTOPHE Henry (BEL)   
1       MACIAS Jose (ARG)  MATEUCCI Francisco (URU)   
2     TEJADA Anibal (URU)   VALLARINO Ricardo (URU)   
3   WARNKEN Alberto (CHI)       LANGENUS Jean (BEL)   
4     REGO Gilberto (BRA)      SAUCEDO Ulises (BOL)   

                  Assistant 2 RoundID  MatchID  Home Team Initials  \
0         REGO Gilberto (BRA)   201.0   1096.0                 FRA   
1       WARNKEN Alberto (CHI)   201.0   1090.0                 USA   
2         BALWAY Thomas (FRA)   201.0   1093.0                 YUG   
3    MATEUCCI Francisco (URU)   201.0   1098.0                 ROU   
4  RADULESCU Constantin (ROU)   201.0   1085.0                 ARG   

  Away Team Initials     Ganador  
0                MEX      France  
1                BEL         USA  
2                BRA  Yugoslavia  
3                PER     Romania  
4                FRA   Argentina  

[5 rows x 21 columns]

In [168]:
(
    df_partidos
    .dropna(subset=["Ganador"])
    .groupby("Ganador")
    .count()
    .sort_values("Year", ascending=False)
)

Year  Datetime  Stage  Stadium  City  \
Ganador                                                            
Brazil                        71        71     71       71    71   
Italy                         45        45     45       45    45   
Argentina                     44        44     44       44    44   
Germany FR                    36        36     36       36    36   
Germany                       32        32     32       32    32   
...                          ...       ...    ...      ...   ...   
Cuba                           1         1      1        1     1   
Korea DPR                      1         1      1        1     1   
Wales                          1         1      1        1     1   
rn">Bosnia and Herzegovina     1         1      1        1     1   
Iran                           1         1      1        1     1   

                            Home Team Name  Home Team Goals  Away Team Goals  \
Ganador                                                                        
Brazil                                  71               71               71   
Italy                                   45               45               45   
Argentina                               44               44               44   
Germany FR                              36               36               36   
Germany                                 32               32               32   
...                                    ...              ...              ...   
Cuba                                     1                1                1   
Korea DPR                                1                1                1   
Wales                                    1                1                1   
rn">Bosnia and Herzegovina               1                1                1   
Iran                                     1                1                1   

                            Away Team Name  Win conditions  Attendance  \
Ganador                                                                  
Brazil                                  71              71          71   
Italy                                   45              45          45   
Argentina                               44              44          44   
Germany FR                              36              36          36   
Germany                                 32              32          32   
...                                    ...             ...         ...   
Cuba                                     1               1           1   
Korea DPR                                1               1           1   
Wales                                    1               1           1   
rn">Bosnia and Herzegovina               1               1           1   
Iran                                     1               1           1   

                            Half-time Home Goals  Half-time Away Goals  \
Ganador                                                                  
Brazil                                        71                    71   
Italy                                         45                    45   
Argentina                                     44                    44   
Germany FR                                    36                    36   
Germany                                       32                    32   
...                                          ...                   ...   
Cuba                                           1                     1   
Korea DPR                                      1                     1   
Wales                                          1                     1   
rn">Bosnia and Herzegovina                     1                     1   
Iran                                           1                     1   

                            Referee  Assistant 1  Assistant 2  RoundID  \
Ganador                                                                  
Brazil                      

Ahora calcularemos para cada país el ratio de victorias, calculado como el número de partidos ganados dividido por el número de partidos jugados.

In [164]:
df_partidos["MatchID"].value_counts()

300186497.0    2
300186502.0    2
300186491.0    2
300186462.0    2
300186485.0    2
              ..
2170.0         1
2067.0         1
2065.0         1
2064.0         1
564.0          1
Name: MatchID, Length: 649, dtype: int64

Podemos observar que tenemos un total de 649 partidos. A continuación calcularemos el número de partidos por país y el número de paises ganados.

In [165]:
total_partidos = df_paises["Country"].value_counts().sort_values(ascending = False)
print(total_partidos)


Brazil                        108
Italy                          83
Argentina                      81
Germany FR                     62
England                        62
                             ... 
Haiti                           3
rn">Bosnia and Herzegovina      3
Israel                          3
Iraq                            3
Dutch East Indies               1
Name: Country, Length: 83, dtype: int64


In [166]:
total_ganados = df_partidos["Ganador"].value_counts().sort_values(ascending = False)
print(total_ganados)

Brazil                        71
Italy                         45
Argentina                     44
Germany FR                    36
Germany                       32
                              ..
Jamaica                        1
Wales                          1
Korea DPR                      1
Tunisia                        1
rn">Bosnia and Herzegovina     1
Name: Ganador, Length: 64, dtype: int64


In [169]:
# Ordenar por ratio
victorias = (
    df_partidos
    .dropna(subset=["Ganador"])
    .groupby("Ganador")
    .count()
    .rename(columns={"Year":"Victorias"})
    [["Victorias"]]
)

jugados = df_paises.groupby("Country").count().rename(columns={"Year": "Jugados"})[["Jugados"]]
ratio_victorias = pd.merge(jugados, victorias, left_index=True, right_index=True, how="inner")
ratio_victorias["Ratio"] = ratio_victorias["Victorias"] / ratio_victorias["Jugados"]

ratio_victorias.sort_values("Ratio", ascending=False)

Jugados  Victorias     Ratio
Germany                       46         32  0.695652
Brazil                       108         71  0.657407
Germany FR                    62         36  0.580645
Argentina                     81         44  0.543210
Italy                         83         45  0.542169
...                          ...        ...       ...
Morocco                       13          2  0.153846
rn">Republic of Ireland       13          2  0.153846
Korea DPR                      7          1  0.142857
Bulgaria                      26          3  0.115385
Tunisia                       12          1  0.083333

[64 rows x 3 columns]